### PR TITLE
[HIP] Document func ptr and virtual func

### PR DESCRIPTION
**Pull request by yxsamliu**
Document clang support for function pointers and virtual functions with HIP

### CODE REVIEW

File: clang/docs/HIPSupport.rst [hunk 1]
```diff
@@ -176,3 +176,70 @@ Predefined Macros
...
+   __global__ void kernel() {
+      Derived obj;
+      Base* basePtr = &obj;
+      basePtr->virtualFunction(); // Allowed since obj is constructed in device code
+   }
+
+Note
+----
+
+Ensure to construct objects in the appropriate context (host or device) depending on where you intend to call their virtual functions to avoid runtime errors.
```
[the remark this answers is not in the document]
Comment (yxsamliu) [reply clang/docs/HIPSupport.rst:245]: will remove

**Review by b-sumner — COMMENTED** [review]
This looks fine to me.

File: clang/docs/HIPSupport.rst [hunk 2]
```diff
@@ -176,3 +176,65 @@ Predefined Macros
...
+     - ``-fno-gpu-rdc`` Mode (default)
+     - ``-fgpu-rdc`` Mode
+   * - Defined and used in the same TU
+     - Supported
+     - Supported
+   * - Defined in one TU and used in another TU
+     - Not Supported
+     - Supported
+
+In the ``-fno-gpu-rdc`` mode, the compiler calculates the resource usage of kernels based only on functions present within the same Translation Unit (TU). This mode does not support the use of function pointers defined in a different TU due to the possibility of incorrect resource usage calculations, leading to undefined behavior.
```
Comment (Artem-B) [inline clang/docs/HIPSupport.rst:198]: I'd mention that in `-fno-gpu-rdc` each TU compiles to a fully linked GPU executable, vs an object file with `-fgpu-rdc`. This should give more context for reasoning about accessibility of various data bits.

File: clang/docs/HIPSupport.rst [hunk 3]
```diff
@@ -176,3 +176,65 @@ Predefined Macros
...
+
+On the other hand, the ``-fgpu-rdc`` mode allows the definition and use of function pointers across different TUs, as resource usage calculations can accommodate functions from disparate TUs.
+
+Virtual Function Support in Clang with HIP
+==========================================
+
+In Clang with HIP, support for calling virtual functions of an object in device or host code is contingent on where the object is constructed.
```
Comment (Artem-B) [inline clang/docs/HIPSupport.rst:205]: We should also mention how mixing host and device virtual functions is expected to work.

AFAICT, vtable populates wrong-side virtual functions with nullptr: https://godbolt.org/z/4f6hvjMcj

File: clang/docs/HIPSupport.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (yxsamliu) [reply clang/docs/HIPSupport.rst]: I added a separate section about -fgpu-rdc and -fno-gpu-rdc mode

File: clang/docs/HIPSupport.rst (hunk 3 above)
Comment (yxsamliu) [reply clang/docs/HIPSupport.rst:205]: will do

File: clang/docs/HIPSupport.rst [hunk 4]
```diff
@@ -176,3 +176,95 @@ Predefined Macros
...
+
+A possible way to alleviate the current limitation of virtual function support in HIP is through the use of a "composite vtable". This involves creating a vtable that combines those from the host and all offloading device architectures, storing it in memory accessible by both. A dedicated registration function is introduced to populate this composite vtable. This function is invoked during global initialization to ensure the vtable is ready before any virtual function calls are made. For every virtual function call, irrespective of context, the system refers to this composite vtable to determine the correct function execution.
```
Comment (scchan) [inline clang/docs/HIPSupport.rst:245]: Could you clarify whether the purpose of this pragraph is to provide a potential workaround to users or a solution in a future compiler?

Comment (yxsamliu) [reply clang/docs/HIPSupport.rst:245]: It is a possible solution. However, there may be better solutions that I am open to discussion.

**Comment by yxsamliu** [comment]
ping

File: clang/docs/HIPSupport.rst (hunk 4 above)
Comment (b-sumner) [reply clang/docs/HIPSupport.rst:245]: I would avoid speculating in this documentation.  Maybe we should remove this paragraph?

Comment (yxsamliu) [reply clang/docs/HIPSupport.rst]: removed the paragraph

**Review by b-sumner — COMMENTED** [review]
Looks good to me.

**Comment by scchan** [comment]
LGTM thanks

**Review by Artem-B — APPROVED** [review]

**Pull request merged**